okio-files/src/commonMain/kotlin/okio/Path.kt

```diff
@@ -0,0 +1,152 @@
+ *
+ * Paths may be absolute or relative:
+ *
+ * * **Absolute paths** are prefixed with `/` and identify a location independent of any working
+ *   directory.
+ * * **Relative paths** are not prefixed with `/`. On their own, relative paths do not identify a
+ *   location on a filesystem; they must be resolved against an absolute path first. When a relative
+ *   path is used to access a [Filesystem], it is resolved against [Filesystem.cwd].
+ *
+ * After the optional leading `/`, the rest of the path is a sequence of segments separated by `/`
+ * characters. Segments satisfy these rules:
+ *
+ *  * Segments are always non-empty.
+ *  * Segments may not be `.`.
+ *  * Segments with the value `..` only occur in relative paths, where these must precede all other
+ *    segments.
+ *
+ * The only path that ends with `/` is the filesystem root, `/`. The empty path `""` is a relative
+ * path that resolves to whichever path it is resolved against.
+ *
+ * Sample Paths
+ * ------------
+ *
+ * | Path                   | Type       | Parent              |
+ * | :--------------------- | :--------- | :------------------ |
+ * | `/Users/jessewilson`   | Absolute   | `/Users`            |
+ * | `/Users`               | Absolute   | `/`                 |
+ * | `/`                    | Absolute   | null                |
+ * | `src/main/java`        | Relative   | `src/main`          |
+ * | `src/main`             | Relative   | `src`               |
+ * | `src`                  | Relative   | (empty)             |
+ * | (empty)                | Relative   | null                |
+ * | `../../src/main/java`  | Relative   | ` ../../src/main`   |
+ * | `../../src/main`       | Relative   | ` ../../src`        |
+ * | `../../src`            | Relative   | ` ../..`            |
+ * | `../..`                | Relative   | null                |
+ */
+class Path private constructor(
+  private val bytes: ByteString
+) {
+  val isAbsolute: Boolean
+    get() = bytes.startsWith(SLASH)
+
+  val isRelative: Boolean
+    get() = !isAbsolute
+
+  /**
+   * Returns the path immediately enclosing this path. This returns null if this is either the
+   * filesystem root (`/`), the empty path, or a traversal up from the empty path.
+   */
+  val parent: Path?
+    get() {
+      if (bytes == EMPTY || bytes == SLASH || bytes.endsWith(SLASH_DOT_DOT) || bytes == DOT_DOT) {
+        return null // Terminal path.
+      }
+      return when (val lastSlash = bytes.lastIndexOf(SLASH)) {
+        -1 -> Path(EMPTY) // Parent is the current working directory.
+        0 -> Path(bytes.substring(endIndex = 1)) // Parent is the filesystem root '/'.
+        else -> Path(bytes.substring(endIndex = lastSlash))
+      }
+    }
+
+  /** Returns a path that resolves [child] relative to this path. */
+  operator fun div(child: String): Path {
+    val buffer = Buffer()
+    buffer.write(bytes)
+    if (buffer.size > 0) {
+      buffer.write(SLASH)
+    }
+    buffer.writeUtf8(child)
```

Comment:
How do I resolve children that are not valid UTF-8?

Reply:
The internal representation as a `ByteString` is my acknowledging that paths aren’t strings. Gotta add more APIs to support byte strings in more places!

Reply:
Yeah i was happy to see that. If you haven't, https://fasterthanli.me/articles/i-want-off-mr-golangs-wild-ride is a fun read.